### ISSUE TITLE
What will happen, if the access token expired in between the Shopify API call (GraphQL/Rest)

### ISSUE THREAD

**Issue by knsakib** [nprutan/shopify-cookieless-auth-demo#4]
I would like to know what will happen, if the token expired in between the Shopify API call.

Let's say, the user started the app(by clicking the app button in the Shopify app section) with a valid access token. And while he was using the app(clicking different sections in the app) the online token expired in the middle of app use.
What will happen then. Will the user be redirected to auth again or not. I did not see that code. I can see redirects in the beginning. For example when a user clicks the app icon with an expired token at that time he will reinitialize the auth and that is fine. But I am concern about expiration in between use. Maybe I am missing some fundamental understanding of how session and access token expiration.

**Comment by nprutan** [comment]
Hi @knsakib, thanks for opening the issue. This is a good question, and it would probably depend on what you are doing on the frontend. I am using "offline" tokens in this example, as I do with most of the apps I build as well, and they don't have an expiration issue. If you're going to use online tokens, see this article (point 2 in the article):
https://shopify.dev/tutorials/authenticate-your-app-using-session-tokens

For this example repo, If the user loads up the base route, the app will try to verify the validity of the token. If it's expired, they will get redirected back to auth. See the verifyToken call here from server.js:

```
router.get("/", async (ctx, next) => {
    const shop = getQueryKey(ctx, "shop");
    // Retrieve token here
    const token = "persistedToken";
    // We are setting the state here explicity
    // for uniformity with ctx.state in createShopifyAuth method
    // State with this shape is used in 
    // koa-shopify-auth-cookieless/utilites/getShopCredentials
    ctx.state = { shopify: { shop: shop, accessToken: token } };
    **await verifyToken(ctx, next);**
  });
```

Now, in your situation, since this is based on Next.js and they might leave the app open with an "online" token, there is the possibility that the token could expire as the window was left open and the user was navigating purely on the frontend with the "next/router". In this case, any GQL calls would fail since they would need a valid token -- not only a valid API token, but also the session token. Either one of these may expire independently. It looks like the expiration timeframe on the session token is 1 minute:
https://shopify.dev/concepts/apps/building-embedded-apps-using-session-tokens

App Bridge should take care of rotating these session tokens, so that shouldn't be an issue. I'm not sure what happens when / if the online token expires and App Bridge tries to rotate the session token. I'm assuming it should redirect, but I haven't tested this. It might be in the docs somewhere.

What I would recommend is creating a hook for your frontend that tracks the status of the token (essentially by proxy) through polling some simple GQL query like {shop {name}}. If the data is not returned, force a redirect.

You might not even need a hook, just do something like this:
```
const GET_SHOP_NAME = gql(`query {shop {name} }`);
const {data, error} = useQuery(GET_SHOP_NAME, {pollInterval: 60000});
useEffect(() => {
    if (data && data.shop.name) {
      console.log("all is good");
    } else {
      redirect.dispatch(Redirect.Action.APP, '/');
    }
  });
```
[](https://shopify.dev/tools/app-bridge/actions/navigation/redirect)

You could also create a route in Koa and ping the route, using the same logic and redirect if there is a failure returned from the route.

See here for help performing a REST call from the backend:
https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/rest.md
Here for GQL from backend:
https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/graphql.md

NOTE: Some of the above mechanisms will probably not work with this repo. I would recommend downloading the Shopify App-CLI and starting from there if you want to use the two mechanisms above. They may work with this repo, but I haven't tested.
https://shopify.dev/tools/cli

Hope that helps!

**Comment by knsakib** [comment]
Thank you. We are using your library https://github.com/nprutan/koa-shopify-auth-cookieless as it has much cleaner implementation than latest https://github.com/Shopify/koa-shopify-auth cookieless implementation. We have a backend route where we do not call GraphQL, so we are verifying session token only. But when we need to call /graphql we need to also verify access token as we want to avoid that 'app window left open issue' or 'possibility of access token expiration' between call for smooth user interaction and make sure the call went through to /graphql (if needed after re-auth because of token expire). But during app use, between user clicks, 'base url' is not there. So just redirect to /auth will not work. But your suggestion of using useEffect and redirect.dispatch seems like a solution. We will come back and let you if that works and solve the issue. Thank you very much for the help.

**Comment by nprutan** [comment]
Great, glad you are finding the package useful! Sounds good, let me know if the polling / redirect works. BTW, I noticed an error in my example above. Obviously redirect to the base route rather than "/settings". I just pulled that from the docs. Otherwise, I have actually used this technique in a production app for redirecting to auth whenever someone bails out of the install process on the scopes page. This was required by Shopify to get the app approved.
